randovania/generator/dock_weakness_distributor.py

```diff
@@ -0,0 +1,193 @@
+import asyncio
+from random import Random
+import time
+from typing import Callable, Tuple
+from randovania.game_description import default_database
+from randovania.game_description.assignment import DockWeaknessAssignment
+from randovania.game_description.game_description import GameDescription
+from randovania.game_description.game_patches import GamePatches
+from randovania.game_description.requirements import Requirement, ResourceRequirement
+from randovania.game_description.resources.item_resource_info import ItemResourceInfo
+from randovania.game_description.resources.resource_type import ResourceType
+from randovania.game_description.resources.simple_resource_info import SimpleResourceInfo
+from randovania.game_description.world.dock_node import DockNode
+from randovania.game_description.world.node import NodeContext
+from randovania.game_description.world.node_identifier import NodeIdentifier
+from randovania.generator.filler.runner import FillerPlayerResult
+from randovania.layout.base.base_configuration import BaseConfiguration
+from randovania.layout.base.dock_rando_configuration import DockRandoMode
+from randovania.lib import random_lib
+from randovania.resolver import debug, resolver
+from randovania.resolver.logic import Logic
+from randovania.resolver.resolver_reach import ResolverReach
+from randovania.resolver.state import State
+
+
+def distribute_pre_fill_weaknesses(patches: GamePatches):
+    dock_rando = patches.configuration.dock_rando
+
+    if dock_rando.mode == DockRandoMode.VANILLA:
+        return patches
+    
+    game = default_database.game_description_for(patches.configuration.game)
+    weakness_database = game.dock_weakness_database
+
+    docks_to_unlock = {
+        node.identifier: weakness_database.dock_rando_params[node.dock_type].unlocked
+        for node in game.world_list.all_nodes
+        if (
+            isinstance(node, DockNode) and dock_rando.types_state[node.dock_type].can_shuffle
+            and node.default_dock_weakness in dock_rando.types_state[node.dock_type].can_change_from
+        )
+    }
+
+    return patches.assign_dock_weakness_assignment(docks_to_unlock)
+
+    
+
+class DockRandoLogic(Logic):
+    dock: DockNode
+
+    def __init__(self, game: GameDescription, configuration: BaseConfiguration, dock: DockNode):
+        super().__init__(game, configuration)
+        self.dock = dock
+    
+    @classmethod
+    def from_logic(cls, logic: Logic, dock: DockNode) -> "DockRandoLogic":
+        return cls(logic.game, logic.configuration, dock)
+
+    @property
+    def victory_condition(self) -> Requirement:
+        return ResourceRequirement(self.dock.identifier, 1, False)
+
+
+TO_SHUFFLE_PROPORTION = 0.6
+RESOLVER_ATTEMPTS = 125
+async def distribute_post_fill_weaknesses(rng: Random,
```

Comment:
Some comments overall to explain what each section intends to do would be helpful. Maybe split off into different functions with docstrings. For example, the setup/calling the resolver could've been a separated function, as well as creating the `weighted_weaknesses`.